### PR TITLE
Remove api versions service

### LINKED ISSUE #1378: Remove ApiVersionsService
Outcome of #641 was a deprecation of the class. It is now eligible for deletion as it will be more than 3 minor releases since 0.3.0

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Refactoring

### Description

Closes #1378

Remove ApiVersionsService
    
Deprecated since 0.3.0 and now eligible for destruction by our policy.
    
    We haven't used it yet and the utility of it to Filters is unclear. The
    original rationale was that Filters can hijack the connection and use
    sendRequest to send messages to the broker, messages constructed by the
    Filter. We need to ensure we only use fields that the broker
    understands, so by offering an API to access the supported versions a
    Filter could discover what's available and react to that.
    
    For now we will remove the deprecated class and can re-implement
    it if anyone needs this functionality. For now the Filters can work
    around it by intercepting the ApiVersionsResponse themselves which
    should work for the typical case where the Filter wants to do some work
    after the api versions are negotiated.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by robobario** [comment]
Given that we haven't had any complaints about it being deprecated with no replacement, should we just remove the feature of giving Filters access to supported versions.

**Comment by k-wall** [comment]
Could we restate the use-case(s) for the API?  What problems are we actually trying to solve?

**Comment by robobario** [comment]
The original rationale from #586 is:

> Filters can choose to hijack the connection to send their own extra requests to the upstream broker. To do this safely we need to know what versions of the RPC the upstream supports, so we don't send it unsupported requests.
>
> Filters are free to implement ApiVersionsResponseFilter and intercept the response, or alternatively send an additional request to the upstream broker asking for the ApiVersions (some clients may not send an ApiVersionsRequest, therefore there is no response to intercept). We can implement this pattern for Filter Authors so it does not need to be implemented in each Filter that wants to know the upstream API versions.
>
> The API provides the intersected versions supported by both upstream and this Kroxylicious instance and also the raw upstream versions in case they are of interest.

**Comment by robobario** [comment]
but given no usage has arrived we could remove it for now and add it when someone needs it

**Comment by k-wall** [comment]
> Filters can choose to hijack the connection to send their own extra requests to the upstream broker. To do this safely we need to know what versions of the RPC the upstream supports, so we don't send it unsupported requests.

I think that model is problematic from the perspective of the broker.

* It means the Broker might be required to hold two conversations at different API versions at once for the same RPC.   I don't think the Kafka protocol actually specifies whether this use-case is valid or not (I doubt it is even thought about).   It might work with Apache Kafka, but who knows about other implementations of the Kafka protocol (Red Panda, Azure etc).  Even if it works now, I think we'd be on thin ice.
* Broker's expose the connection metadata information as metrics (client software/version).  This API has the possibility that broker receives two different values.  How's the broker to behave?
  
I also question whether that API as specified is actually useful.  As a Filter Author, I might want to create/cache a response that I send back to the client as a response over and over.  To do this, I need the knowledge of what downstream supports in order to create that response.  At the moment this API doesn't help me achieve that.

So, I'm wondering if the API should be simply:

```
 CompletionStage<Optional<ApiVersionsResponseData.ApiVersion>> getLastNegotiatedApiVersion(ApiKeys keys)
```

which would return the last return a result from the **last 3-way API version intersection** between client/proxy/broker.  If the client hasn't sent a `API_VERSION` request (yet?), then the Optional will be empty.  This will be good for most filters that are interested in filtering the 'business' RPCs after occur the connection opening preamble is done.

**Comment by robobario** [comment]
>   As a Filter Author, I might want to create/cache a response that I send back to the client as a response over and over. To do this, I need the knowledge of what downstream supports in order to create that response at the right version. At the moment this API doesn't help me achieve that.

The response will always be to a request that has a specific apiVersion from the client, right? So I don't think we need to know the 3-way intersection to generate responses, because the request carries the apiVersion we need to respond with.

**Comment by k-wall** [comment]
> > As a Filter Author, I might want to create/cache a response that I send back to the client as a response over and over. To do this, I need the knowledge of what downstream supports in order to create that response at the right version. At the moment this API doesn't help me achieve that.
> 
> The response will always be to a request that has a specific apiVersion from the client, right? So I don't think we need to know the 3-way intersection to generate responses, because the request carries the apiVersion we need to respond with.

That's true. Maybe my use-case of creating responses before hand was a bit contrived.

**Comment by k-wall** [comment]
> but given no usage has arrived we could remove it for now and add it when someone needs it

+1

**Comment by robobario** [comment]
I have now removed it entirely

**Comment by robobario** [comment]
hey, sonar found some dead code, thanks sonar

File: kroxylicious-integration-tests/pom.xml [hunk 1]
```diff
@@ -151,17 +151,6 @@
             <artifactId>testcontainers</artifactId>
             <scope>test</scope>
         </dependency>
-        <!-- the next two are used by the Netty Leak Detection test -->
```
Comment (robobario) [inline kroxylicious-integration-tests/pom.xml:154]: found this dead code while I was cleaning imports

**Review by SamBarker — APPROVED** [review]
LGTM

**Pull request merged**